File: java-adman-spring/application/src/main/java/com/example/adman/BusinessPartnerController.java
package com.example.adman;

import com.google.gson.Gson;
import com.sap.cloud.sdk.s4hana.datamodel.odata.namespaces.businesspartner.BusinessPartner;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@RestController
public class BusinessPartnerController {

    private static final Logger logger =
            LoggerFactory.getLogger(BusinessPartnerController.class);

    @GetMapping( "/api/business-partners" )
    ResponseEntity<String> doGet(
            @RequestParam(required = false) String id)  {

        String jsonResult;
        if (id == null) {
            // Get all Business Partners
            List<BusinessPartner> list;
            try {
                list = BusinessPartnerRepository.findAll();
            }
            catch (BusinessPartnerRepository_InternalErrorException e)
            {
                final String message = "Internal error " +
                        "retrieving business partners.";
                logger.error(message, e);
                throw new ResponseStatusException(
                        HttpStatus.INTERNAL_SERVER_ERROR, message, e);
            }

            // Convert to JSon and remove null values
            jsonResult = new Gson().toJson(list);
        }
        else {
            // Get Business Partner based on id
            BusinessPartner bp;
            try {
                bp = BusinessPartnerRepository.findById(id);
            }
            catch (BusinessPartnerRepository_InternalErrorException e)
            {
                final String message
                        = "Internal error retrieving " +
                        "business partner " + id;
                logger.error(message, e);
                throw new ResponseStatusException(
                        HttpStatus.INTERNAL_SERVER_ERROR, message, e);
            }

            // Convert to JSon and remove null values
            jsonResult = new Gson().toJson(bp);
        }

        // Prepare response headers (optional)
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        return new ResponseEntity<>(
                jsonResult, headers, HttpStatus.OK);
    }
}

File: ts-adman/src/adman/business-partner.controller.ts
import { Controller, Get, Query } from '@nestjs/common';
import { BusinessPartnerService } from './business-partner.service';
import { BusinessPartner } from '@sap/cloud-sdk-vdm-business-partner-service';

@Controller('api/business-partners')
export class BusinessPartnerController {

  @Get()
  async getBusinessPartners(@Query('id') id: string)
    : Promise<BusinessPartner[] | BusinessPartner> {
    if (id == null) {
      return BusinessPartnerService.findAll()
        .catch(error => {
          console.log(`Failed to get business partners - ${error.message}`);
          return null;
        });
    }
    else {
      return BusinessPartnerService.findById(id)
      .catch(error => {
        console.log(`Failed to get business partner ${id} - ${error.message}`);
        return null;
      });
    }
  }
}
